rfcs/0024-hard-forks.md

```diff
@@ -0,0 +1,455 @@
+  type era_id_2 = (true_ty,  true_ty)  t
+
+  type existential =
+    | Era0 : era_id_0 -> existential
+    | Era1 : era_id_1 -> existential
+    | Era2 : era_id_2 -> existential
+end
+
+let incr_count : (true_ty, 'a) Block.t -> (true_ty, 'a) Block.t = fun b ->
+  {b with transaction_counter =
+    Static_option.map ~f:(fun x -> x + 1) b.transaction_counter
+  }
+
+(** Update the parallel scan state, dispatching on which one is turned on. A
+    separate function would handle generating a block that is an era transition.
+*)
+let update_pscan :
+  type use_new_pscan.
+  ('htc, use_new_pscan) Block.t -> ('htc, use_new_pscan) Block.t = fun b ->
+   match b.pscan with
+   (* The type system prevents us from calling Old_pscan.update on a
+      block with the new data structure and vice versa. *)
+     | SLeft ps -> Old_pscan.update b
+     | SRight ps -> New_pscan.update b
+```
+
+Blocks that come from or are sent to the network will be (de)serialized via
+`Block.existential`. This scheme will unfortunately require writing some
+serialization code manually or modifying the `bin_prot` ppx.
+
+To keep the number of type parameters small and mitigate the combinatorial
+explosion of possible configurations, as time passes we'll remove type
+parameters that are set to the same value in all currently supported eras (now +
+the next one) and change the types of the associated fields to directly include
+the values.
+
+#### Handling changes in types
+
+Different kinds of changes will have to be implemented differently. If types
+from a new era only add optional fields, or new fields that have a reasonable
+default, then we can convert incoming data from era n into the era n+1 version,
+and only store the items with the type parameterization of the current era. If
+this isn't possible then we will have to either have both options in an
+`Either.t` or in some cases two parallel data structures.
+
+##### Blocks + block components
+
+If there is no or a small enough change to the consensus mechanism then the
+transition frontier can have blocks of both types, or blocks of only the new
+type with conversion. A sufficiently different consensus mechanism might need a
+separate transition frontier or other data structure to support it. In that
+case, blocks from era n will go in the old transition frontier and blocks from
+era n+1 will go into the new structure. The era transition block will be a
+special case.
+
+##### User transactions
+
+As of now our rule is that we only accept transactions into the pool if they
+could be applied to the best tip. Continuing in this vein, we will only accept
+transactions into the pool if they are valid in the current era. This will cause
+a small hiccup in throughput when era transitions happen, but that's acceptable
+since they are rare.
+
+##### Ledger SNARKs
+
+We also do not accept ledger SNARKs until we've entered the appropriate era for
+them and there is a block whose children might need to buy them.
```

Comment:
So what happens to the ledger proofs that are generated during the parallel scan transition phase?